### ISSUE TITLE
Develop the ability to compare pdfs to each other

### ISSUE THREAD

**Issue by plocket** [SuffolkLITLab/ALKiln#25]
The eventual goal would be a step to compare an existing pdf to a downloaded pdf. First we have to figure out how to compare two PDFs, though.

**Comment by plocket** [comment]
Look into https://github.com/jcushman/pdfquery

**Comment by rpigneri-vol** [comment]
Or check nodejs.com.

**Comment by rpigneri-vol** [comment]
I'm going to evaluate the following libraries:

- https://www.npmjs.com/package/@exzeo/pdf-diff
- https://www.npmjs.com/package/pdf-diff
- https://www.npmjs.com/package/compare-pdf (I might skip this one because it has some ungainly dependencies).

**Comment by rpigneri-vol** [comment]
https://github.com/jcushman/pdfquery is in Python so this is disqualified.

**Comment by rpigneri-vol** [comment]
Option 1 does not seem to recognize filled values.  In other words, it only compares the static parts of the PDF and completely ignores any filled values.  However, it was updated two months ago and might still be actively developed so it might be worth it to fill out an issue if no other project can check these values.

**Comment by rpigneri-vol** [comment]
Option 2 converts the PDF's to JSON and diffs the JSON.  The dependent libraries explicitly support the contents of checkboxes and fillable forms.  However, the interface to this library is a little tough to use.  Here's a text summary of the diff object returned when the library compares a PDF to itself:

[
  '~',
  {
    Fields: [
      [Array], [Array],
      [Array], [Array],
      [Array], [Array],
      [Array], [Array],
      [Array], [Array],
      [Array]
    ],
    Boxsets: [
      [Array], [Array],
      [Array], [Array],
      [Array], [Array],
      [Array], [Array],
      [Array], [Array]
    ]
  }
]

Here's the output when the two files are different:

[
  '~',
  {
    Fields: [
      [Array], [Array],
      [Array], [Array],
      [Array], [Array],
      [Array], [Array],
      [Array], [Array],
      [Array]
    ],
    Boxsets: [
      [Array], [Array],
      [Array], [Array],
      [Array], [Array],
      [Array], [Array],
      [Array], [Array]
    ]
  }
]

This could do the trick, but the interface would take some getting used to and need some cleanup wrappers.

**Comment by rpigneri-vol** [comment]
For option 3, we have confirmed that GitHub can install ImageMagick via part of the workflow (https://stackoverflow.com/questions/67060829/github-action-using-imagemagick-create-a-png-and-commit-it-to-the-repo).  Therefore, if ImageMagick really is a dependency, we can install it outside of npm.

We also believe that we might be able to use compare-pdf without the external (i.e, non-Node) dependencies.  We'll try this out in the next session.

**Comment by rpigneri-vol** [comment]
Option #3 appears to work and does not actually require ImageMagick, GraphicsMagick, and GhostScript.  The interface is a little delicate, but it appears that once it is mastered, the code is fairly stable.  It does check for fillable form values.

I would say that option #3 does appear to be the winner.

**Comment by rpigneri-vol** [comment]
When I downloaded the code separately and ran it, I got the same errors as when I run my own tests.  I suspect that a dependency got updated and now the library doesn't work.

**Comment by rpigneri-vol** [comment]
Let's find something in Node before we consider stuff in Python (or rewriting the Python comparator in Node).

Let's try to find some PDF's whose static portions are different.  First, let's establish a test threshold by using the lowest possible threshold for two identical documents.  Then, use that on two different PDF's.

Next steps:
1. Convert to black and white.
2. Normalize colors or other attributes.

**Comment by rpigneri-vol** [comment]
We will put this down for now.   When we come back, we'll take a look at the JavaScript or node.js.

**Comment by plocket** [comment]
Using https://emn178.github.io/online-tools/md5_checksum.html with 'File Hash' MD5, this seems to (in at least these cases) see the difference in fillable fields. Ignore the filenames.

File 1 with one field filled out with "test" and "test" and a date:
[Simple_Sample.pdf](https://github.com/SuffolkLITLab/ALKiln/files/9480338/Simple_Sample.pdf)

File 2 with the same fields filled out with the same information, but created in its own version of the file at a different creation time comes out equal to file 1:
[Simple_Sample (2).pdf](https://github.com/SuffolkLITLab/ALKiln/files/9480350/Simple_Sample.2.pdf)

File 3 with different values in the fields comes out unequal to file 1:
[Simple_Sample (1).pdf](https://github.com/SuffolkLITLab/ALKiln/files/9480351/Simple_Sample.1.pdf)

**Comment by plocket** [comment]
There might be a js lib called javascript-md5

**Comment by plocket** [comment]
This will also depend on how da creates these files too.

**Comment by BryceStevenWilley** [comment]
Also could consider: https://mozilla.github.io/pdf.js/getting_started/.  It handles viewing PDFs in Firefox, so should be able to access field info from the PDFs.

**Comment by plocket** [comment]
Using the above methods, there were a couple duplicate documents that each had the same field values that @BryceStevenWilley tried that came out with different hashes. They should have come out as being equal.

**Comment by BryceStevenWilley** [comment]
I got something working in node with pdf.js, on the [pdfjs branch](https://github.com/SuffolkLITLab/ALKiln/tree/pdfjs). The 20 line script (strongly stolen from [the examples](https://github.com/mozilla/pdf.js/blob/master/examples/node/getinfo.js)) that can get the fields and the full text content of the PDF (for if someone leaves the fields as fillable, or if they permanently bake them into the PDF);

```node
#!/usr/bin/env node

const pdfjsLib = require("pdfjs-dist/legacy/build/pdf.js");

// Loading file from file system into typed array
const pdfPath =
  process.argv[2] || "Adult Name Change (Alabama).pdf";

async function testPdf(pdfPath) {
  const loadingTask = await pdfjsLib.getDocument(pdfPath);

  const doc = await loadingTask.promise;
  const contents = await (await doc.getPage(1)).getTextContent();
  console.log(contents.items.map((item) => item.str).join(" "));
  const field_objects = await doc.getFieldObjects();
  console.log(JSON.stringify(field_objects, null, 4));
  return field_objects;
}

testPdf(pdfPath);
```

We do still need to make the step and the API for this, but I'm confident pdf.js is what we need.

**Comment by BryceStevenWilley** [comment]
Made some more progress on PDF comparison (still on the [pdfjs branch](https://github.com/SuffolkLITLab/ALKiln/tree/pdfjs)). I added a few internal functions to `utils/pdf_reader.js`:

* `get_content`: gets all of the text content from the PDF
* `get_fields`: gets all field objects from the PDF (NOTE: won't be there if people have their PDFs set to `editable: False` or `pdfa: True` in the DA interviews
* `get_field_value`: gets the specific value of a PDF field
* `compare_pdfs`: takes two PDFS and makes sure that the text content is identical, and all of the PDF fields are identical.

I added a single step: 

* `I expect the PDFs at "some/path" and "another/path" to be the same`.

Differently open to feedback on these APIs, if we want to add or modify any.

I still need to test what's there, but once I'm sure it's working, I'll open a PR with it.
